src/R/Components/Impl/Repl/Commands/SourceRScriptCommand.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.Common.Core.Shell;
+using Microsoft.R.Components.ContentTypes;
+using Microsoft.R.Components.Controller;
+using Microsoft.R.Components.Extensions;
+using Microsoft.R.Components.InteractiveWorkflow;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.VisualStudio.R.Package.Repl.Commands {
+    public sealed class SourceRScriptCommand : IMenuCommand {
+        private readonly IRInteractiveWorkflow _interactiveWorkflow;
+        private readonly IActiveWpfTextViewTracker _activeTextViewTracker;
+        private readonly bool _echo;
+
+        public SourceRScriptCommand(
+            IRInteractiveWorkflow interactiveWorkflow,
+            IActiveWpfTextViewTracker activeTextViewTracker,
+            bool echo
+        ) {
+            _interactiveWorkflow = interactiveWorkflow;
+            _activeTextViewTracker = activeTextViewTracker;
+            _echo = echo;
+        }
+
+        public CommandStatus Status {
+            get {
+                var status = CommandStatus.Supported;
+                if (_interactiveWorkflow.ActiveWindow == null) {
+                    status |= CommandStatus.Invisible;
+                } else if (!string.IsNullOrEmpty(GetFilePath())) {
+                    status |= CommandStatus.Enabled;
+                }
+                return status;
+            }
+        }
+
+        public CommandResult Invoke(object inputArg, ref object outputArg) {
```

Comment:
Currently we do something like this (from `ShowHelpOnCurrentCommand`):
```{csharp}
        protected override void Handle() {
            try {
                if (!_workflow.RSession.IsHostRunning) {
                    return;
                }

                // Fetch identifier under the cursor
                string item = GetItemUnderCaret();
                if (item == null || item.Length >= MaxHelpItemLength) {
                    return;
                }

                _workflow.RSession.ExecuteAsync($"rtvs:::show_help({item.ToRStringLiteral()})")
                    .SilenceException<RException>()
                    .SilenceException<MessageTransportException>()
                    .DoNotWait();
            } catch (Exception ex) {
                Debug.Assert(false, string.Format(CultureInfo.InvariantCulture, "Help on current item failed. Exception: {0}", ex.Message));
                // Catch everything so exceptions don't leave the async void method
                if (ex.IsCriticalException()) {
                    throw;
                }
            }
        }
```

With such implementation, there is no good way to wait until command is executed. We can call `DoNotWait` inside CommandShim and make `Handle` async`. This will not change implementation, but will help in test.

---
In reply to: [63559342](https://github.com/Microsoft/RTVS/pull/1785#discussion_r63559342) [](ancestors = 63559342)